packages/@aws-cdk/aws-ec2/lib/vpc-endpoint-service.ts

```diff
@@ -0,0 +1,87 @@
+import { ArnPrincipal } from '@aws-cdk/aws-iam';
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { CfnVPCEndpointService, CfnVPCEndpointServicePermissions } from './ec2.generated';
+
+/**
+ * A VPC endpoint service.
+ */
+export interface IVpcEndpointService extends IResource {
+}
+
+/**
+ * A VPC endpoint service
+ * @resource AWS::EC2::VPCEndpointService
+ */
+export class VpcEndpointService extends Resource implements IVpcEndpointService {
+
+  /**
+   * One or more network load balancer ARNs to host the service.
+   * @attribute
+   */
+  public readonly networkLoadBalancerArns: string[];
+
+  /**
+   * Whether to require manual acceptance of new connections to the service.
+   * @attribute
+   */
+  public readonly acceptanceRequired: boolean;
+
+  /**
+   * One or more Principal ARNs to allow inbound connections to.
+   * @attribute
+   */
+  public readonly whitelistedPrincipals: ArnPrincipal[];
+
+  private readonly endpointService: CfnVPCEndpointService;
+
+  constructor(scope: Construct, id: string, props: VpcEndpointServiceProps) {
+    super(scope, id);
+
+    if (props.networkLoadBalancerArns === undefined || props.networkLoadBalancerArns.length === 0) {
+      throw new Error("VPC Endpoint Service must have at least one `Network Load Balancer ARN` specified.");
+    }
+
+    this.networkLoadBalancerArns = props.networkLoadBalancerArns;
+    this.acceptanceRequired = props.acceptanceRequired !== undefined ? props.acceptanceRequired : true;
+    this.whitelistedPrincipals = props.whitelistedPrincipals !== undefined ? props.whitelistedPrincipals : [];
+
+    this.endpointService = new CfnVPCEndpointService(this, id, {
+      networkLoadBalancerArns: props.networkLoadBalancerArns,
+      acceptanceRequired: this.acceptanceRequired
+    });
+
+    new CfnVPCEndpointServicePermissions(this, id + "Permissions", {
+      serviceId: this.endpointService.ref,
+      allowedPrincipals: this.whitelistedPrincipals.map(x => x.arn)
+    });
+  }
+}
+
+/**
+ * Construction properties for a VpcEndpointService.
+ */
+export interface VpcEndpointServiceProps {
+
+    /**
+     * Name of the Vpc Endpoint Service
+     * @default - CDK generated name
+     */
+    readonly vpcEndpointServiceName?: string;
+
+    /**
+     * One or more network load balancer ARNs to host the service.
+     */
+    readonly networkLoadBalancerArns: string[];
```

Comment:
Is this a String because of the cyclic dependency problem? Otherwise I would have this be an array of like ILoadBalancer

Reply:
Yes :(

Reply:
I would ask @rix0rrr if there’s some fancy way around this problem. For context, we can’t have ‘real’ load balancers here because the ELB module depends on EC2 and adding them here would create a cyclical dependency.

Reply:
Yep, this is not cool.

Since the ELB modules already depend on the EC2 modules, how about we do the following:

We define an interface here, that we take:

```ts
export interface IVpcEndpointLoadBalancer {
  public readonly loadBalancerArn: string;
}
```

We make `INetworkLoadBalancer` extends this interface:

```ts
export interface INetworkLoadBalancer extends IVpcEndpointLoadBalancer {
  public readonly loadBalancerArn: string;   // <-- was already there if everything is correct
```

We can now use `NetworkLoadBalancer` directly as an argument to `VpcServiceEndpoint`, and we've also given ourselves an extension point if they ever start support any other kind of resource.

Reply:
Oh dear. I ran into the same problem when writing unit tests for VpcEndpointService, because I cannot import NetworkLoadBalancer due to a circular dependency. Any recommendations? Would a dummy IVpcEndpointServiceLoadBalancer be appropriate?

Reply:
Added this

Reply:
New invocation:
```
VpcEndpointService = new VpcEndpointService(this, "Endpoint Service", VpcEndpointServiceProps.builder()
                .vpcEndpointServiceLoadBalancers(ArrayList<NetworkLoadBalancer>())
                .acceptanceRequired(false)
                .allowedPrincipals(ArrayList<ArnPrincipal>())
                .build());
```